src/RelayKnownVariableNames.ts

```diff
@@ -0,0 +1,66 @@
+import { ValidationRule } from "graphql"
+import { GraphQLError } from "./dependencies"
+import { getArgumentDefinitions } from "./argumentDefinitions"
+
+export function undefinedVarMessage(varName: string, opName?: string): string {
+  return opName
+    ? `Variable "$${varName}" is not defined by operation "${opName}".`
+    : `Variable "$${varName}" is not defined.`
+}
+
+export function undefinedVarMessageFragment(varName: string, fragmentName: string): string {
+  return `Variable "$${varName}" is not defined by fragment "${fragmentName}".`
+}
+
+export const RelayKnownVariableNames: ValidationRule = function RelayKnownVariableNames(context) {
+  let variableNameDefined = Object.create(null)
+  return {
+    OperationDefinition: {
+      enter() {
+        variableNameDefined = Object.create(null)
+      },
+      leave(operation) {
+        const usages = context.getVariableUsages(operation)
+
+        usages.forEach(({ node }) => {
+          const varName = node.name.value
+          if (variableNameDefined[varName] !== true) {
```

Comment:
Nit: can this just be `!variableNameDefined[varName]` or is this a situation where you don’t want to compare against falsy values other than `false`?

Reply:
I think I actually just copied this from the main validation done by graphql - I'm fine with changing it :)

Reply:
Hah, fair enough. Thanks, tho!